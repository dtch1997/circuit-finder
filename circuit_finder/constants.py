import torch
import pathlib
from circuit_finder.core.types import LayerIndex

GPT_2_SMALL: str = "gpt2"
ALL_GPT_2_SMALL_LAYERS: list[LayerIndex] = list(range(12))

device = "cuda" if torch.cuda.is_available() else "cpu"

ProjectDir = pathlib.Path(__file__).parent.parent
