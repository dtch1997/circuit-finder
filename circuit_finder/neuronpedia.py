import json
import urllib.parse

from circuit_finder.core.types import SaeFamily


def get_neuronpedia_url_for_quick_list(
    layer: int,
    features: list[int],
    sae_family: SaeFamily = "res-jb",
    name: str = "temporary_list",
):
    url = "https://neuronpedia.org/quick-list/"
    name = urllib.parse.quote(name)
    url = url + "?name=" + name
    list_feature = [
        {
            "modelId": "gpt2-small",
            "layer": f"{layer}-{sae_family}",
            "index": str(feature),
        }
        for feature in features
    ]
    url = url + "&features=" + urllib.parse.quote(json.dumps(list_feature))
    return url
