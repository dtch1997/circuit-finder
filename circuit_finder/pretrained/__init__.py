from .load_attn_saes import load_attn_saes
from .load_mlp_transcoders import load_mlp_transcoders
from .load_resid_saes import load_resid_saes
