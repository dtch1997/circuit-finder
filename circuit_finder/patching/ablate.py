from einops import einsum, rearrange
import torch
from jaxtyping import Int, Float
from functools import partial
from circuit_finder.core.types import MetricFn, Tokens
from circuit_finder.patching.eap_graph import EAPGraph
from torch import Tensor
import transformer_lens as tl

from transcoders_slim.transcoder import Transcoder
from circuit_finder.pretrained.load_mlp_transcoders import ts_tc_to_hooked_tc
from circuit_finder.core.hooked_sae import HookedSAE
from circuit_finder.core.hooked_transcoder import HookedTranscoder
from circuit_finder.core.hooked_transcoder import HookedTranscoderReplacementContext


def get_mask(
    graph, module_name, layer, feature_acts: Float[Tensor, "batch seq n_features"]
):
    """
    Returns mask : [seq, n_features].
    mask[seq, feature_id] is False whenever {module_name}.{layer}.{seq}.{feature_id} is a node of graph.
    All other entries are True
    """
    nodes = graph.get_src_nodes()
    positions = [
        int(node.split(".")[2])
        for node in nodes
        if node.split(".")[:2] == [module_name, str(layer)]
    ]
    feature_ids = [
        int(node.split(".")[3])
        for node in nodes
        if node.split(".")[:2] == [module_name, str(layer)]
    ]
    mask = torch.ones_like(feature_acts[0]).bool()
    mask[positions, feature_ids] = False
    return mask


def get_metric_with_ablation(
    model: tl.HookedTransformer,
    graph: EAPGraph,
    tokens: Int[Tensor, "batch seq"],
    metric: MetricFn,
    transcoders,  # list
    attn_saes,  # list
    ablate_nodes: str | bool = "zero",  # options [False, "bm", "zero"]
    ablate_errors: str | bool = False,  # options [False, "bm", "zero"]
    first_ablated_layer: int = 2,  # Marks et al don't ablate first 2 layers
    freeze_attention: bool = False,
):
    model = add_ablation_hooks_to_model(
        model,
        graph,
        tokens,
        transcoders,
        attn_saes,
        ablate_nodes,
        ablate_errors,
        first_ablated_layer,
        freeze_attention,
    )
    return metric(model, tokens)


def get_forward_cache(
    model: tl.HookedSAETransformer,
    tokens: Tokens,
    transcoders: list[Transcoder] | list[HookedTranscoder],
    attn_saes: list[HookedSAE],
) -> tl.ActivationCache:
    hooked_transcoders = []
    for tc in transcoders:
        if isinstance(tc, Transcoder):
            tc = ts_tc_to_hooked_tc(tc)
        hooked_transcoders.append(tc)

    # Run model
    with HookedTranscoderReplacementContext(
        model,  # type: ignore
        transcoders=hooked_transcoders,
    ) as context:
        with model.saes(saes=attn_saes):  # type: ignore
            _, cache = model.run_with_cache(tokens)

    return cache


def add_ablation_hooks_to_model(
    model: tl.HookedTransformer,
    graph: EAPGraph,
    tokens: Int[Tensor, "batch seq"],
    transcoders,  # list
    attn_saes,  # list
    ablate_nodes: str | bool = "zero",  # options [False, "bm", "zero"]
    ablate_errors: str | bool = False,  # options [False, "bm", "zero"]
    first_ablated_layer: int = 2,  # Marks et al don't ablate first 2 layers
    freeze_attention: bool = False,
) -> tl.HookedTransformer:
    """Cache the activations of the model on a circuit, then return the model with relevant ablation hooks added"""
    assert ablate_errors in [False, "bm", "zero"]
    assert ablate_nodes in [False, "bm", "zero"]
    # Do clean FP to get batchmean (bm) feature acts, and reconstructions
    # No patching is done during this FP, so the hook fns return nothing
    # Note: MLP transcoders are a bit more fiddly than attn-SAEs, requiring us to cache at both mlp_in and mlp_out
    mlp_bm_feature_act_cache = {}
    attn_bm_feature_act_cache = {}
    mlp_recons_cache = {}
    mlp_error_cache = {}
    attn_error_cache = {}

    def mlp_in_cache_hook(act, hook, layer):
        assert hook.name.endswith("ln2.hook_normalized")
        t = transcoders[layer]
        feature_acts = torch.relu(
            einsum(
                act - t.b_dec,
                t.W_enc,
                "b s d_model, d_model n_features -> b s n_features",
            )
            + t.b_enc
        )
        recons = feature_acts @ t.W_dec + t.b_dec_out

        mlp_bm_feature_act_cache[layer] = feature_acts.mean(0)
        mlp_recons_cache[layer] = recons

    def mlp_out_cache_hook(act, hook, layer):
        assert hook.name.endswith("mlp_out")
        mlp_error_cache[layer] = act - mlp_recons_cache[layer]

    def attn_cache_hook(act, hook, layer):
        assert hook.name.endswith("hook_z")
        sae = attn_saes[layer]
        z_concat = rearrange(act, "b s n_heads d_head -> b s (n_heads d_head)")
        feature_acts = torch.relu(
            einsum(
                z_concat - sae.b_dec,
                sae.W_enc,
                "b s d_model, d_model n_features -> b s n_features",
            )
            + sae.b_enc
        )
        recons_concat = feature_acts @ sae.W_dec + sae.b_dec
        recons = rearrange(
            recons_concat,
            "b s (n_heads d_head) -> b s n_heads d_head",
            n_heads=model.cfg.n_heads,
        )

        attn_bm_feature_act_cache[layer] = feature_acts.mean(0)
        attn_error_cache[layer] = act - recons

    model.reset_hooks()
    for layer in range(model.cfg.n_layers):
        model.add_hook(
            f"blocks.{layer}.ln2.hook_normalized",
            partial(mlp_in_cache_hook, layer=layer),
            "fwd",
        )
        model.add_hook(
            f"blocks.{layer}.hook_mlp_out",
            partial(mlp_out_cache_hook, layer=layer),
            "fwd",
        )
        model.add_hook(
            f"blocks.{layer}.attn.hook_z", partial(attn_cache_hook, layer=layer), "fwd"
        )

    # Run forward pass and populate all caches
    _, pattern_cache = model.run_with_cache(
        tokens, return_type="loss", names_filter=lambda x: x.endswith("pattern")
    )
    assert len(pattern_cache) > 0
    assert len(mlp_bm_feature_act_cache) > 0
    assert len(attn_bm_feature_act_cache) > 0
    assert len(mlp_recons_cache) > 0
    assert len(mlp_error_cache) > 0
    assert len(attn_error_cache) > 0

    # Now do FP where we patch nodes not in graph with their batchmeaned values
    model.reset_hooks()
    mlp_ablated_recons = {}

    def freeze_pattern_hook(act, hook):
        assert hook.name.endswith("pattern")
        return pattern_cache[hook.name]

    def mlp_out_ablated_recons_cache_hook(act, hook, layer):
        assert hook.name.endswith("ln2.hook_normalized")
        t = transcoders[layer]
        feature_acts = torch.relu(
            einsum(
                act - t.b_dec,
                t.W_enc,
                "b s d_model, d_model n_features -> b s n_features",
            )
            + t.b_enc
        )
        mask = get_mask(graph, "mlp", layer, feature_acts)
        ablated_feature_acts = feature_acts.clone()
        ablated_feature_acts[:, mask] = mlp_bm_feature_act_cache[layer][mask]
        mlp_ablated_recons[layer] = ablated_feature_acts @ t.W_dec + t.b_dec_out

    def mlp_patch_hook(act, hook, layer):
        assert hook.name.endswith("mlp_out")

        if ablate_nodes == "bm":
            recons = mlp_ablated_recons[layer]
        elif ablate_nodes == "zero":
            recons = torch.zeros_like(mlp_ablated_recons[layer])
        else:
            # NOTE: This requires us to change the way we're computing the cached objects.
            raise NotImplementedError

        if ablate_errors == "bm":
            error = mlp_error_cache[layer].mean(dim=0, keepdim=True)
        elif ablate_errors == "zero":
            error = torch.zeros_like(mlp_error_cache[layer])
        else:
            error = mlp_error_cache[layer]

        return recons + error

    def attn_patch_hook(act, hook, layer):
        assert hook.name.endswith("hook_z")
        sae = attn_saes[layer]
        z_concat = rearrange(act, "b s n_heads d_head -> b s (n_heads d_head)")
        feature_acts = torch.relu(
            einsum(
                z_concat - sae.b_dec,
                sae.W_enc,
                "b s d_model, d_model n_features -> b s n_features",
            )
            + sae.b_enc
        )

        mask = get_mask(graph, "attn", layer, feature_acts)
        ablated_feature_acts = feature_acts.clone()
        ablated_feature_acts[:, mask] = attn_bm_feature_act_cache[layer][mask]
        ablated_recons_concat = ablated_feature_acts @ sae.W_dec + sae.b_dec
        ablated_recons = rearrange(
            ablated_recons_concat,
            "b s (n_heads d_head) -> b s n_heads d_head",
            n_heads=model.cfg.n_heads,
        )

        if ablate_nodes == "bm":
            recons = ablated_recons
        elif ablate_nodes == "zero":
            recons = torch.zeros_like(ablated_recons)
        else:  # no ablation
            raise NotImplementedError

        if ablate_errors == "bm":
            error = attn_error_cache[layer].mean(dim=0, keepdim=True)
        elif ablate_errors == "zero":
            error = torch.zeros_like(attn_error_cache[layer])
        else:
            error = attn_error_cache[layer]

        return recons + error

    for layer in range(first_ablated_layer, model.cfg.n_layers):
        model.add_hook(
            f"blocks.{layer}.ln2.hook_normalized",
            partial(mlp_out_ablated_recons_cache_hook, layer=layer),
            "fwd",
        )
        model.add_hook(
            f"blocks.{layer}.hook_mlp_out", partial(mlp_patch_hook, layer=layer), "fwd"
        )
        model.add_hook(
            f"blocks.{layer}.attn.hook_z", partial(attn_patch_hook, layer=layer), "fwd"
        )
        if freeze_attention:
            model.add_hook(
                f"blocks.{layer}.attn.hook_pattern", freeze_pattern_hook, "fwd"
            )

    return model
