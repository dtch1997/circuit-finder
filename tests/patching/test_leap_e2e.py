"""Check that the LEAP algorithm returns a given circuit"""

import pytest
import transformer_lens as tl

from circuit_finder.pretrained import (
    load_attn_saes,
    load_mlp_transcoders,
)
from circuit_finder.patching.leap import (
    preprocess_attn_saes,
    LEAP,
    LEAPConfig,
)

from circuit_finder.constants import device


@pytest.mark.skip
def test_leap_e2e(snapshot):
    # Load models
    model = tl.HookedTransformer.from_pretrained(
        "gpt2",
        device=device,
        fold_ln=True,
        center_writing_weights=True,
        center_unembed=True,
    )

    attn_saes = load_attn_saes()
    attn_saes = preprocess_attn_saes(attn_saes, model)  # type: ignore
    transcoders = load_mlp_transcoders()

    # Load dataset
    tokens = model.to_tokens(
        ["When John and Mary were at the store, John gave a bottle to Mary"]
    )
    corrupt_tokens = model.to_tokens(
        ["When Alice and Bob were at the store, Alice gave a bottle to Bob"]
    )

    # Set up LEAP
    cfg = LEAPConfig(threshold=0.2, contrast_pairs=True)
    leap = LEAP(
        cfg, tokens, model, attn_saes, transcoders, corrupt_tokens=corrupt_tokens
    )

    # Run LEAP
    leap.metric_step()
    for layer in reversed(range(1, model.cfg.n_layers)):
        leap.mlp_step(layer)
        leap.ov_step(layer)

    graph = leap.graph
    assert graph == snapshot
