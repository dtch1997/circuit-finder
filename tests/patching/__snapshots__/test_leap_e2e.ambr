# serializer version: 1
# name: test_leap_e2e
  list([
    tuple(
      tuple(
        'null',
        'metric.12.14.0',
      ),
      0,
    ),
    tuple(
      tuple(
        'metric.12.14.0',
        'attn.9.14.11368',
      ),
      5.972317218780518,
    ),
    tuple(
      tuple(
        'metric.12.14.0',
        'attn.10.14.3849',
      ),
      1.9210439920425415,
    ),
    tuple(
      tuple(
        'attn.10.14.3849',
        'mlp.0.4.5545',
      ),
      11.198338508605957,
    ),
    tuple(
      tuple(
        'attn.10.14.3849',
        'mlp.1.4.12820',
      ),
      0.2502943277359009,
    ),
    tuple(
      tuple(
        'attn.10.14.3849',
        'mlp.1.4.20449',
      ),
      1.5138646364212036,
    ),
    tuple(
      tuple(
        'attn.10.14.3849',
        'mlp.2.4.21793',
      ),
      1.277372121810913,
    ),
    tuple(
      tuple(
        'attn.10.14.3849',
        'mlp.3.4.3104',
      ),
      0.26564618945121765,
    ),
    tuple(
      tuple(
        'attn.10.14.3849',
        'mlp.3.4.11805',
      ),
      0.28001999855041504,
    ),
    tuple(
      tuple(
        'attn.10.14.3849',
        'mlp.3.4.18695',
      ),
      0.2829084098339081,
    ),
    tuple(
      tuple(
        'attn.10.14.3849',
        'mlp.4.4.12781',
      ),
      0.4094651937484741,
    ),
    tuple(
      tuple(
        'attn.10.14.3849',
        'mlp.5.4.6255',
      ),
      1.547159194946289,
    ),
    tuple(
      tuple(
        'attn.10.14.3849',
        'mlp.6.4.22620',
      ),
      1.2372771501541138,
    ),
    tuple(
      tuple(
        'attn.10.14.3849',
        'mlp.7.4.12771',
      ),
      1.9184175729751587,
    ),
    tuple(
      tuple(
        'attn.10.14.3849',
        'mlp.8.4.3750',
      ),
      1.2165870666503906,
    ),
    tuple(
      tuple(
        'attn.10.14.3849',
        'mlp.9.4.13544',
      ),
      0.7327587604522705,
    ),
    tuple(
      tuple(
        'attn.10.14.3849',
        'attn.0.4.5263',
      ),
      1.7582728862762451,
    ),
    tuple(
      tuple(
        'attn.10.14.3849',
        'attn.9.14.11368',
      ),
      0.3921414613723755,
    ),
    tuple(
      tuple(
        'mlp.9.4.13544',
        'mlp.0.4.5545',
      ),
      5.413640022277832,
    ),
    tuple(
      tuple(
        'mlp.9.4.13544',
        'mlp.1.4.20449',
      ),
      0.7689730525016785,
    ),
    tuple(
      tuple(
        'mlp.9.4.13544',
        'mlp.2.4.21793',
      ),
      0.5978395342826843,
    ),
    tuple(
      tuple(
        'mlp.9.4.13544',
        'mlp.4.4.12781',
      ),
      0.3410624861717224,
    ),
    tuple(
      tuple(
        'mlp.9.4.13544',
        'mlp.5.4.6255',
      ),
      0.8134390115737915,
    ),
    tuple(
      tuple(
        'mlp.9.4.13544',
        'mlp.6.4.22620',
      ),
      0.8883530497550964,
    ),
    tuple(
      tuple(
        'mlp.9.4.13544',
        'mlp.7.4.12771',
      ),
      1.524917721748352,
    ),
    tuple(
      tuple(
        'mlp.9.4.13544',
        'mlp.8.4.3750',
      ),
      1.4344134330749512,
    ),
    tuple(
      tuple(
        'mlp.9.4.13544',
        'attn.0.4.5263',
      ),
      1.1035138368606567,
    ),
    tuple(
      tuple(
        'attn.9.14.11368',
        'mlp.0.2.13881',
      ),
      0.5740739107131958,
    ),
    tuple(
      tuple(
        'attn.9.14.11368',
        'mlp.1.2.5913',
      ),
      0.2944994270801544,
    ),
    tuple(
      tuple(
        'attn.9.14.11368',
        'mlp.0.4.5545',
      ),
      23.875911712646484,
    ),
    tuple(
      tuple(
        'attn.9.14.11368',
        'mlp.1.4.20449',
      ),
      2.974609375,
    ),
    tuple(
      tuple(
        'attn.9.14.11368',
        'mlp.2.4.21793',
      ),
      3.375497817993164,
    ),
    tuple(
      tuple(
        'attn.9.14.11368',
        'mlp.3.4.11805',
      ),
      0.9724463820457458,
    ),
    tuple(
      tuple(
        'attn.9.14.11368',
        'mlp.3.4.13400',
      ),
      0.32081934809684753,
    ),
    tuple(
      tuple(
        'attn.9.14.11368',
        'mlp.3.4.18695',
      ),
      0.8792752623558044,
    ),
    tuple(
      tuple(
        'attn.9.14.11368',
        'mlp.4.4.12781',
      ),
      0.763002872467041,
    ),
    tuple(
      tuple(
        'attn.9.14.11368',
        'mlp.5.4.645',
      ),
      0.3336610794067383,
    ),
    tuple(
      tuple(
        'attn.9.14.11368',
        'mlp.5.4.3821',
      ),
      0.31864026188850403,
    ),
    tuple(
      tuple(
        'attn.9.14.11368',
        'mlp.5.4.6255',
      ),
      2.930612564086914,
    ),
    tuple(
      tuple(
        'attn.9.14.11368',
        'mlp.5.4.9019',
      ),
      0.22704516351222992,
    ),
    tuple(
      tuple(
        'attn.9.14.11368',
        'mlp.6.4.16515',
      ),
      0.3843565583229065,
    ),
    tuple(
      tuple(
        'attn.9.14.11368',
        'mlp.6.4.22620',
      ),
      2.695824146270752,
    ),
    tuple(
      tuple(
        'attn.9.14.11368',
        'mlp.7.4.2739',
      ),
      0.3912883996963501,
    ),
    tuple(
      tuple(
        'attn.9.14.11368',
        'mlp.7.4.12771',
      ),
      3.561506509780884,
    ),
    tuple(
      tuple(
        'attn.9.14.11368',
        'mlp.8.4.3750',
      ),
      2.0000481605529785,
    ),
    tuple(
      tuple(
        'attn.9.14.11368',
        'mlp.8.4.23192',
      ),
      0.2303709238767624,
    ),
    tuple(
      tuple(
        'attn.9.14.11368',
        'mlp.0.10.13881',
      ),
      0.2924053370952606,
    ),
    tuple(
      tuple(
        'attn.9.14.11368',
        'attn.0.4.5263',
      ),
      5.884767532348633,
    ),
    tuple(
      tuple(
        'attn.9.14.11368',
        'attn.1.4.4734',
      ),
      0.24821272492408752,
    ),
    tuple(
      tuple(
        'mlp.8.4.23192',
        'mlp.0.4.5545',
      ),
      1.697969675064087,
    ),
    tuple(
      tuple(
        'mlp.8.4.23192',
        'mlp.1.4.20449',
      ),
      0.2888564467430115,
    ),
    tuple(
      tuple(
        'mlp.8.4.3750',
        'mlp.0.4.5545',
      ),
      5.364632606506348,
    ),
    tuple(
      tuple(
        'mlp.8.4.3750',
        'mlp.1.4.12820',
      ),
      0.20498467981815338,
    ),
    tuple(
      tuple(
        'mlp.8.4.3750',
        'mlp.1.4.20449',
      ),
      0.9824649095535278,
    ),
    tuple(
      tuple(
        'mlp.8.4.3750',
        'mlp.2.4.21793',
      ),
      0.6743975877761841,
    ),
    tuple(
      tuple(
        'mlp.8.4.3750',
        'mlp.3.4.11805',
      ),
      0.20581594109535217,
    ),
    tuple(
      tuple(
        'mlp.8.4.3750',
        'mlp.4.4.12781',
      ),
      0.30503934621810913,
    ),
    tuple(
      tuple(
        'mlp.8.4.3750',
        'mlp.5.4.6255',
      ),
      0.8138896226882935,
    ),
    tuple(
      tuple(
        'mlp.8.4.3750',
        'mlp.6.4.22620',
      ),
      0.9841546416282654,
    ),
    tuple(
      tuple(
        'mlp.8.4.3750',
        'mlp.7.4.12771',
      ),
      2.030116558074951,
    ),
    tuple(
      tuple(
        'mlp.8.4.23192',
        'attn.0.4.5263',
      ),
      0.6092767715454102,
    ),
    tuple(
      tuple(
        'mlp.8.4.3750',
        'attn.0.4.5263',
      ),
      1.4698492288589478,
    ),
    tuple(
      tuple(
        'mlp.7.4.12771',
        'mlp.0.4.5545',
      ),
      6.746464252471924,
    ),
    tuple(
      tuple(
        'mlp.7.4.12771',
        'mlp.1.4.20449',
      ),
      0.9597537517547607,
    ),
    tuple(
      tuple(
        'mlp.7.4.12771',
        'mlp.2.4.21793',
      ),
      0.8283870220184326,
    ),
    tuple(
      tuple(
        'mlp.7.4.12771',
        'mlp.3.4.3104',
      ),
      0.25551679730415344,
    ),
    tuple(
      tuple(
        'mlp.7.4.12771',
        'mlp.3.4.11805',
      ),
      0.29598355293273926,
    ),
    tuple(
      tuple(
        'mlp.7.4.12771',
        'mlp.3.4.18695',
      ),
      0.21180209517478943,
    ),
    tuple(
      tuple(
        'mlp.7.4.12771',
        'mlp.4.4.12781',
      ),
      0.46139073371887207,
    ),
    tuple(
      tuple(
        'mlp.7.4.12771',
        'mlp.5.4.6255',
      ),
      1.0643324851989746,
    ),
    tuple(
      tuple(
        'mlp.7.4.12771',
        'mlp.6.4.22620',
      ),
      1.2137845754623413,
    ),
    tuple(
      tuple(
        'mlp.7.4.2739',
        'mlp.0.4.5545',
      ),
      2.263742685317993,
    ),
    tuple(
      tuple(
        'mlp.7.4.2739',
        'mlp.1.4.20449',
      ),
      0.23725615441799164,
    ),
    tuple(
      tuple(
        'mlp.7.4.2739',
        'mlp.2.4.21793',
      ),
      0.2846706509590149,
    ),
    tuple(
      tuple(
        'mlp.7.4.2739',
        'mlp.5.4.6255',
      ),
      0.2762969136238098,
    ),
    tuple(
      tuple(
        'mlp.7.4.2739',
        'mlp.6.4.22620',
      ),
      0.548466145992279,
    ),
    tuple(
      tuple(
        'mlp.7.4.12771',
        'attn.0.4.5263',
      ),
      1.817074179649353,
    ),
    tuple(
      tuple(
        'mlp.7.4.2739',
        'attn.0.4.5263',
      ),
      1.0170390605926514,
    ),
    tuple(
      tuple(
        'mlp.6.4.22620',
        'mlp.0.4.5545',
      ),
      5.028758525848389,
    ),
    tuple(
      tuple(
        'mlp.6.4.22620',
        'mlp.1.4.12820',
      ),
      0.25564298033714294,
    ),
    tuple(
      tuple(
        'mlp.6.4.22620',
        'mlp.1.4.20449',
      ),
      0.7481616735458374,
    ),
    tuple(
      tuple(
        'mlp.6.4.22620',
        'mlp.2.4.21793',
      ),
      0.7778315544128418,
    ),
    tuple(
      tuple(
        'mlp.6.4.22620',
        'mlp.3.4.11805',
      ),
      0.25532424449920654,
    ),
    tuple(
      tuple(
        'mlp.6.4.22620',
        'mlp.4.4.12781',
      ),
      0.47690680623054504,
    ),
    tuple(
      tuple(
        'mlp.6.4.22620',
        'mlp.5.4.6255',
      ),
      0.9449648857116699,
    ),
    tuple(
      tuple(
        'mlp.6.4.16515',
        'mlp.0.4.5545',
      ),
      1.12620210647583,
    ),
    tuple(
      tuple(
        'mlp.6.4.16515',
        'mlp.0.4.9713',
      ),
      0.49409323930740356,
    ),
    tuple(
      tuple(
        'mlp.6.4.16515',
        'mlp.2.4.21793',
      ),
      0.36152708530426025,
    ),
    tuple(
      tuple(
        'mlp.6.4.22620',
        'attn.0.4.5263',
      ),
      1.3158185482025146,
    ),
    tuple(
      tuple(
        'mlp.6.4.16515',
        'attn.0.4.5263',
      ),
      0.21511709690093994,
    ),
    tuple(
      tuple(
        'mlp.5.4.6255',
        'mlp.0.4.5545',
      ),
      4.764146327972412,
    ),
    tuple(
      tuple(
        'mlp.5.4.6255',
        'mlp.1.4.12820',
      ),
      0.2106454223394394,
    ),
    tuple(
      tuple(
        'mlp.5.4.6255',
        'mlp.1.4.20449',
      ),
      0.5707271099090576,
    ),
    tuple(
      tuple(
        'mlp.5.4.6255',
        'mlp.2.4.21793',
      ),
      0.5589013695716858,
    ),
    tuple(
      tuple(
        'mlp.5.4.6255',
        'mlp.3.4.3104',
      ),
      0.308554083108902,
    ),
    tuple(
      tuple(
        'mlp.5.4.6255',
        'mlp.4.4.12781',
      ),
      0.4794498085975647,
    ),
    tuple(
      tuple(
        'mlp.5.4.9019',
        'mlp.0.4.5545',
      ),
      3.683314323425293,
    ),
    tuple(
      tuple(
        'mlp.5.4.9019',
        'mlp.1.4.20449',
      ),
      0.2165897786617279,
    ),
    tuple(
      tuple(
        'mlp.5.4.9019',
        'mlp.3.4.3104',
      ),
      0.6001410484313965,
    ),
    tuple(
      tuple(
        'mlp.5.4.3821',
        'mlp.0.4.5545',
      ),
      2.230759382247925,
    ),
    tuple(
      tuple(
        'mlp.5.4.3821',
        'mlp.1.4.20449',
      ),
      0.28243327140808105,
    ),
    tuple(
      tuple(
        'mlp.5.4.645',
        'mlp.0.4.5545',
      ),
      2.1160144805908203,
    ),
    tuple(
      tuple(
        'mlp.5.4.645',
        'mlp.2.4.21793',
      ),
      0.7645828127861023,
    ),
    tuple(
      tuple(
        'mlp.5.4.6255',
        'attn.0.4.5263',
      ),
      1.2950184345245361,
    ),
    tuple(
      tuple(
        'mlp.5.4.9019',
        'attn.0.4.5263',
      ),
      0.46802932024002075,
    ),
    tuple(
      tuple(
        'mlp.5.4.3821',
        'attn.0.4.5263',
      ),
      1.018750786781311,
    ),
    tuple(
      tuple(
        'mlp.5.4.645',
        'attn.0.4.5263',
      ),
      0.77689129114151,
    ),
    tuple(
      tuple(
        'mlp.4.4.12781',
        'mlp.0.4.5545',
      ),
      4.008573532104492,
    ),
    tuple(
      tuple(
        'mlp.4.4.12781',
        'mlp.1.4.12820',
      ),
      0.2553254961967468,
    ),
    tuple(
      tuple(
        'mlp.4.4.12781',
        'mlp.1.4.20449',
      ),
      0.49133777618408203,
    ),
    tuple(
      tuple(
        'mlp.4.4.12781',
        'mlp.2.4.21793',
      ),
      0.574211835861206,
    ),
    tuple(
      tuple(
        'mlp.4.4.12781',
        'mlp.3.4.11805',
      ),
      0.30270230770111084,
    ),
    tuple(
      tuple(
        'mlp.4.4.12781',
        'attn.0.4.5263',
      ),
      1.2326014041900635,
    ),
    tuple(
      tuple(
        'mlp.3.4.3104',
        'mlp.0.4.5545',
      ),
      5.25502347946167,
    ),
    tuple(
      tuple(
        'mlp.3.4.3104',
        'mlp.1.4.12820',
      ),
      0.6672856211662292,
    ),
    tuple(
      tuple(
        'mlp.3.4.3104',
        'mlp.2.4.21793',
      ),
      0.30600956082344055,
    ),
    tuple(
      tuple(
        'mlp.3.4.13400',
        'mlp.0.4.5545',
      ),
      2.32938551902771,
    ),
    tuple(
      tuple(
        'mlp.3.4.13400',
        'mlp.1.4.12820',
      ),
      0.4876801371574402,
    ),
    tuple(
      tuple(
        'mlp.3.4.13400',
        'mlp.1.4.20449',
      ),
      0.49683237075805664,
    ),
    tuple(
      tuple(
        'mlp.3.4.11805',
        'mlp.0.4.5545',
      ),
      4.377124786376953,
    ),
    tuple(
      tuple(
        'mlp.3.4.11805',
        'mlp.1.4.12820',
      ),
      0.2519930601119995,
    ),
    tuple(
      tuple(
        'mlp.3.4.11805',
        'mlp.1.4.20449',
      ),
      0.21329182386398315,
    ),
    tuple(
      tuple(
        'mlp.3.4.18695',
        'mlp.0.4.5545',
      ),
      2.9143271446228027,
    ),
    tuple(
      tuple(
        'mlp.3.4.18695',
        'mlp.0.4.9713',
      ),
      0.23941472172737122,
    ),
    tuple(
      tuple(
        'mlp.3.4.18695',
        'mlp.2.4.21793',
      ),
      0.9226467609405518,
    ),
    tuple(
      tuple(
        'mlp.3.4.3104',
        'attn.0.4.5263',
      ),
      1.274143099784851,
    ),
    tuple(
      tuple(
        'mlp.3.4.13400',
        'attn.0.4.5263',
      ),
      0.6482621431350708,
    ),
    tuple(
      tuple(
        'mlp.3.4.11805',
        'attn.0.4.5263',
      ),
      0.9416947364807129,
    ),
    tuple(
      tuple(
        'mlp.3.4.18695',
        'attn.0.4.5263',
      ),
      0.7635356783866882,
    ),
    tuple(
      tuple(
        'mlp.3.4.18695',
        'attn.3.4.18319',
      ),
      0.20414146780967712,
    ),
    tuple(
      tuple(
        'attn.3.4.18319',
        'mlp.0.2.13881',
      ),
      2.550184965133667,
    ),
    tuple(
      tuple(
        'attn.3.4.18319',
        'mlp.1.2.5913',
      ),
      1.203685998916626,
    ),
    tuple(
      tuple(
        'attn.3.4.18319',
        'mlp.2.2.7897',
      ),
      0.6606404781341553,
    ),
    tuple(
      tuple(
        'attn.3.4.18319',
        'mlp.0.4.5545',
      ),
      0.2275557667016983,
    ),
    tuple(
      tuple(
        'attn.3.4.18319',
        'attn.0.2.9715',
      ),
      0.3100516200065613,
    ),
    tuple(
      tuple(
        'mlp.2.2.7897',
        'mlp.0.2.13881',
      ),
      7.364142417907715,
    ),
    tuple(
      tuple(
        'mlp.2.2.7897',
        'mlp.1.2.5913',
      ),
      3.0168685913085938,
    ),
    tuple(
      tuple(
        'mlp.2.4.21793',
        'mlp.0.4.5545',
      ),
      7.135644912719727,
    ),
    tuple(
      tuple(
        'mlp.2.4.21793',
        'mlp.1.4.12820',
      ),
      0.32562416791915894,
    ),
    tuple(
      tuple(
        'mlp.2.4.21793',
        'mlp.1.4.20449',
      ),
      1.2190444469451904,
    ),
    tuple(
      tuple(
        'mlp.2.2.7897',
        'attn.0.2.9715',
      ),
      0.8858076333999634,
    ),
    tuple(
      tuple(
        'mlp.2.4.21793',
        'attn.0.4.5263',
      ),
      1.4249728918075562,
    ),
    tuple(
      tuple(
        'mlp.1.2.5913',
        'mlp.0.2.13881',
      ),
      11.924360275268555,
    ),
    tuple(
      tuple(
        'mlp.1.4.20449',
        'mlp.0.4.5545',
      ),
      8.602848052978516,
    ),
    tuple(
      tuple(
        'mlp.1.4.12820',
        'mlp.0.4.5545',
      ),
      5.140222072601318,
    ),
    tuple(
      tuple(
        'mlp.1.2.5913',
        'attn.0.2.9715',
      ),
      1.7663376331329346,
    ),
    tuple(
      tuple(
        'mlp.1.4.20449',
        'attn.0.4.5263',
      ),
      2.282473087310791,
    ),
    tuple(
      tuple(
        'mlp.1.4.12820',
        'attn.0.4.5263',
      ),
      1.1159541606903076,
    ),
    tuple(
      tuple(
        'attn.1.4.4734',
        'mlp.0.2.13881',
      ),
      0.30165842175483704,
    ),
    tuple(
      tuple(
        'attn.1.4.4734',
        'mlp.0.4.5545',
      ),
      2.9356179237365723,
    ),
    tuple(
      tuple(
        'attn.1.4.4734',
        'attn.0.4.5263',
      ),
      0.6073545217514038,
    ),
  ])
# ---
